index.js

```diff
@@ -175,40 +228,83 @@
-                    self.listeners.splice(idx, 1);
-            };
-        };
-        EventEmitter.prototype.emit = function(data) {
-            this.listeners.forEach(function (fn) {
-                fn(data);
+            render: function() {
+                return React.Children.only(this.props.children);
+            },
+
+            getChildContext: function () {
+                var stores = {};
+                // inherit stores
+                var baseStores = this.context.mobxStores;
+                if (baseStores) for (var key in baseStores) {
+                    stores[key] = baseStores[key];
+                }
+                // add own stores
+                for (var key in this.props)
+                    if (!specialReactKeys[key])
+                        stores[key] = this.props[key];
+                return {
+                    mobxStores: stores
+                };
+            },
+
+            componentWillReceiveProps: function(nextProps) {
+                // Maybe this warning is to aggressive?
+                if (Object.keys(nextProps).length !== Object.keys(this.props).length)
+                    console.warn("MobX Provider: The set of provided stores has changed. Please avoid changing stores as the change might not propagate to all children");
+                for (var key in nextProps)
+                    if (!specialReactKeys[key] && this.props[key] !== nextProps[key])
+                        console.warn("MobX Provider: Provided store '" + key + "' has changed. Please avoid replacing stores as the change might not propagate to all children");
+            }
+        });
+
+        var PropTypes = React.PropTypes;
+        Provider.contextTypes = { mobxStores: PropTypes.object };
+        Provider.childContextTypes = { mobxStores: PropTypes.object.isRequired };
+
+        /**
+         * Store Injection
+         */
+        function createStoreInjector(stores, component) {
+            var Injector = React.createClass({
+                displayName: "MobXStoreInjector",
+                render: function() {
+                    var newProps = {};
+                    for (var key in this.props)
+                        newProps[key] = this.props[key];
+                    var baseStores = this.context.mobxStores;
+                    stores.forEach(function(storeName) {
+                        if (storeName in newProps) // prefer props over stores
+                            return;
+                        if (!(storeName in baseStores))
+                            throw new Error("MobX observer: Store '" + storeName + "' is not available! Make sure it is provided by some Provider");
+                        newProps[storeName] = baseStores[storeName];
+                    }, this);
+                    return React.createElement(component, newProps);
+                }
             });
-        };
+            Injector.contextTypes = { mobxStores: PropTypes.object.isRequired };
```

Comment:
Should not be required but optional, because it is allowed to provide stores using props instead of context (for example in tests this is useful)